### ISSUE TITLE
Implement a Tracing/APM tool and identify performance hot spots

### ISSUE THREAD

**Issue by bt3gl** [reactioncommerce/reaction#5295]
Implement a tracing or APM tool to the application, and work to identify performance hot spots:

- look at spots where the app interacts with the Mongo database for large collections (e.g. find spots we’re pulling the entire collection from the DB into memory, doing some kind of in-memory sort, and then pushing the result to the client). 

- dig on the performance, focusing on high-impact/traffic queries like our product listing page, navigation queries, product detail, cart, and anything related to checkout.

- possible sub-task: create an enterprise-sized data set that works with the stock-reaction environment (perhaps creating a following up ticket).

**Comment by bt3gl** [comment]
After reading about several available APM tools, I went down to Spencer's initial suggestions:

### New Relic
_Trial: https://newrelic.com/signup?trial=apm_

- Response Time, Throughput, Error Rates etc.
- Performance of external services.
- Most time-consuming transactions.
- Cross-Application Tracing.
- Good database monitoring: SQL query analysis, slow SQL report, time spent in database calls, search all DB/cache operations.
- Good if we wanted to focus on some easy dashboards.

 
### Honeycomb
_https://www.honeycomb.io/_

Tracing Whiteparper: https://www.honeycomb.io/wp-content/uploads/2019/04/Distributed-Tracing-

1. %E2%80%93-A-Guide-for-Microservices-and-More-final.pdf
2. Focused on “distributed tracing”.
3. Accepts events with no schema or index. 
4. It seems very easy to write queries (“break down”, “calculate per group”, “filter”, etc.)
5. Horizontal sharding indefinitely.
6. BubbleUp: seems to be an in-house algorithm for outlier/anomaly detection.
7. New, but very promising. Greate community.
8. Good if we want to take more control of our tracing.

I am now starting testing their trial.

**Comment by spencern** [comment]
Great summary @bt3gl - thanks for the update!

**Comment by kschingiz** [comment]
@bt3gl Hey, maybe you will be interested in Elastic APM, I have developed meteor agent for elastic:
1. APM: https://github.com/kschingiz/meteor-elastic-apm
2. Metrics: https://github.com/kschingiz/meteor-measured

**Comment by janus-reith** [comment]
@kschingiz Interesting, I'm currently investigating aswell what would be the best way of integrating Elastic APM here and also found your package. Will this still give access to the other integrations `elastic-apm-node` offers on default?
Mostly I'm thinking of apollo server and the raw mongo calls that run separate from those that still rely on meteor.

**Comment by kschingiz** [comment]
@janus-reith Yes, meteor-elastic-apm uses internally elastic-apm-node and all its features + I have published new version yesterday which integrates meteor-elastic-apm with meteor-measured which means you can build meteor metrics charts in Kibana, learn how to do it in this [Metrics docs](https://github.com/kschingiz/meteor-elastic-apm/blob/master/METRICS.md)

**Comment by janus-reith** [comment]
That sounds great, will try it out!
